rfcs/rfc-113.md

```diff
@@ -0,0 +1,341 @@
+                "arn:aws:dynamodb:*:*:table/TestTable"
+              ]
+            },
+            {
+              "Action": "logs:PutLogEvents",
+              "Effect": "Allow",
+              "Resource": "*"
+            }
+          ],
+          "Version": "2012-10-17"
+        },
+        "Description": "",
+        "ManagedPolicyName": "TestPolicy",
+        "Path": "/",
+        "Roles": [
+          {
+            "Ref": "TestRole"
+          }
+        ]
+      }
+    }
+  }
+}
+```
+
+Customers expect the rule to filter down to only items of the Statement array which match the given check (the first statement item and not the second in the above example. First has a dynamodb:GetItem action). Instead, the check checks ALL statements, because the given resource satisfied the check. This again violates our tenet of the language being unambiguous, as rule authors are expecting the rule to behave one way when in reality it does not apply as expected.
+
+Rule authors should be able to filter these nested arrays in the same way that they can filter resource types which match a certain check. To accomplish this, we are proposing filtering techniques to select areas of templates that need to be checked.
+
+Example of using filtering with the proposed syntax and new variable assignment from context (explained in depth in the proposal section):
+```
+AWS::IAM::Policy PolicyDocument.Statement[ Action.* == dynamodb:GetItem ].Resource.* != "*"
+```
+
+The above filtering checks only resources which meet the requirement in brackets, then checks said statements for the wildcard resource type check. With the proposals below, we hope to empower ruleset authors to create complex checks with ease, much of which was not possible in the current version of the language.
+
+## **Verbosity when checking deeply nested properties**
+Checking deeply nested properties, like encryption configuration for S3 buckets, the ruleset authoring experience becomes verbose:
+
+To access the configuration, one needs to first author a property access like this:
+```
+AWS::S3::Bucket BucketEncryption.ServerSideEncryptionConfiguration.*.ServerSideEncryptionByDefault
+```
+then make comparisons. If there are more properties that need to be checked in the ServerSideEncryptionByDefault object, this can get verbose quickly. We are proposing enhanced variable assignment from incoming templates. Previously, variables could only be set in the ruleset files themselves and did not have dynamic values. We want to allow customers to do some of these deeply nested checks in an easy fashion by setting a variable to refer to the access, then reusing the variable for a less verbose language.
+
+Assigning parts of the incoming data to variables will allow rule authors to compose simple to read rules:
+```
+%encryption = BucketEncryption.ServerSideEncryptionConfiguration.*.ServerSideEncryptionByDefault
+%encryption.property == "VALUE"
+$encryption.otherProperty IN ["ALLOWED", "VALUES"]
```

Comment:
think I prefer the block-syntax discussed in https://github.com/aws-cloudformation/cloudformation-guard/pull/111 so context can't be split across different parts of rulesets

Reply:
a lot of this was trying to keep the block evaluations out of this PR as to not be adding those changes to the proposal as well